File: scripts/latex/tables/tabular.jl
include("../../commons/utils.jl")
include("alignment.jl")
include("row.jl")

"""
    struct Tabular

A structure for representing a LaTeX tabular environment.

  This structure is used to store the content of a LaTeX tabular environment along with the column 
  alignment specifications.

# Fields
  - `content`: A matrix of any type, representing the contents of the tabular.
    Each cell in the matrix corresponds to a cell in the tabular.
  - `alignment`: A string representing the alignment of the columns.
    The alignment string must contain the same number of alignment specifiers (e.g., "c", "l", "r")
    as there are columns in the content.

# Example
  ```julia
  julia> Tabular([["Name", "Age"], ["John", 32], ["Jane", 28]], "lc")
  Tabular([["Name", "Age"], ["John", 32], ["Jane", 28]], "lc")
  ```
"""	
struct Tabular
  content::Vector{Row}
  alignment::Alignment
end

@doc raw"""
    Base.show(io::IO, ::MIME"text/latex", tab::Tabular)

Render a `Tabular` object to a LaTeX `tabular` environment.

  This function prints the LaTeX code for a `tabular` environment with the content and alignment
  specifications of the given `Tabular` object `tab`.

# Arguments
  - `io`: An `IO` stream to which the LaTeX code is written.
  - `tab`: A `Tabular` object containing the content and alignment for the tabular.

# Exceptions
  - Throws an `ArgumentError` if the number of alignment specifiers does not match 
    the number of columns in `tab.content`.

# Output
  - The LaTeX `tabular` environment is written to the `IO` stream.

# Examples
  ```julia
  julia> t = Tabular([
      "Name" "Age"
      "Alice" 30
      "Bob" 25],
    "ll"
  )
  julia> show(stdout, MIME("text/latex"), t)
  \begin{tabular}{ll}
    Name & Age \\
    \hline
    Alice & 30 \\
    Bob & 25 \\
  \end{tabular}
  ```
"""
function Base.show(io::IO, mime::MIME"text/latex", tab::Tabular)
  println(io, "\\begin{tabular}{$(repr(mime, tab.alignment))}")
  for r in tab.content
    println(io, indent(repr(mime, r), 2))
  end
  print(io, "\\end{tabular}")
end

function get(tab::Tabular, i::Int)::Row
  return tab.content[i]
end

"""
    tabular(content::Matrix; alignment::Alignment

Construct a `Tabular` object with the specified content and alignment.

  The function creates a `Tabular` object that represents a LaTeX tabular environment.
  It takes as input a matrix that specifies the content of the table and an optional string that
  specifies the alignment of each column.
  If no alignment is specified, the default is center alignment for all columns.

# Arguments
  - `content`: A matrix specifying the content of the table.
    Each entry corresponds to a cell in the table.
  - `alignment`: An optional string specifying the alignment of each column.
    The string should contain one character for each column, where "c" indicates center alignment, 
    "l" indicates left alignment, and "r" indicates right alignment.
    If no alignment is specified, the default is center alignment for all columns.

# Returns
  - A `Tabular` object that represents the specified LaTeX tabular environment.

# Examples
  ```julia
  julia> t = tabular(["Alice" "Bob"; 30 25], alignment="ll")
  Tabular(["Alice" "Bob"; 30 25], "ll")
  ```
"""
tabular(
  content::Vector{Row};
  alignment::Alignment
) = Tabular(content, alignment)

File: scripts/latex/tables/tabular.py
from dataclasses import dataclass

from latex.env import environment, MandatoryArgument
from latex.tables.alignments import Alignment
from latex.tables.cells import Cell
from latex.tables.rows import Row


@dataclass
class Tabular:
    """
    Represents a LaTeX tabular environment.

    A Tabular object consists of multiple Row objects and an Alignment object, defining the alignment
    of each column in the tabular.
    """
    rows: tuple[Row, ...]
    alignment: Alignment

    def __init__(self, *content: object, alignment: str | Alignment):
        """
        Construct a new Tabular object.

        :param content: The Row objects to be displayed in the tabular.
                        Each object will be encapsulated into a Row object.
        :param alignment: An Alignment object defining the alignment for each column.
        :raise ValueError: If no content is provided to the tabular.
        """
        if len(content) == 0:
            raise ValueError("Tabular content must not be empty")
        self.rows = tuple([Row(c) for c in content])
        self.alignment = Alignment(alignment)

    def __str__(self):
        """
        Represent the Tabular object as a LaTeX tabular environment string.

        :return: A string representing the Tabular object in the LaTeX tabular environment format.
        """
        return environment("tabular", "\n".join([str(r) for r in self.rows]), MandatoryArgument(str(self.alignment)))

    def __repr__(self):
        """
        Represent the Tabular object as a Python string.

        :return: A string representing the Tabular object in the Python string format.
        """
        return f'Tabular({", ".join([repr(r) for r in self.rows])}, alignment={repr(self.alignment)})'


if __name__ == '__main__':
    print(Tabular(Row(1.0), Row(2.0), alignment=Alignment('r')))
    print(Tabular(1, 2, 3, alignment='r'))
    print(Tabular(Row(1.0, 2.0), Row(3.0, 4.0), alignment=Alignment('r|r')))
    print(Tabular(Row(Cell(1.0, Alignment('r'), 2), 2), Row(3.0, 4.0, 5), alignment=Alignment('r|r')))
